### ISSUE TITLE
User management support

### ISSUE THREAD

**Issue by harshsinghal** [Portkey-AI/gateway#115]
Thanks for open-sourcing the LLM gateway. 

Curious to learn if you are going to add support for user management. 

E.g., feature where I can share API keys with specific users and create allow-list for which LLMs can be accessed by each specific user key. 

Use case could be providing keys to students in a call and allowing them access models on Anyscale and Together. And some students have their keys whitelisted for OpenAI access (per key quota would be great feature).

**Comment by roh26it** [comment]
We have these as virtual keys in the hosted version at Portkey.ai. You should check that out, was built exactly for this purpose. 

https://portkey.ai/docs/product/ai-gateway-streamline-llm-integrations/virtual-keys

We're thinking about open-sourcing some of the middleware we've written for the gateway as well. But we won't be able to offer all the security guarantees, and hence have not done it so far. 

Thanks for the suggestion! Let us know if you have something in mind or if the hosted product works.

**Comment by harshsinghal** [comment]
Got it. I did look into this but maybe I misunderstood the feature.
When I create a virtual key for say Anyscale, I provide my Anyscale API key. I can name the virtual key but this doesn't give me another API key that I can then share with user.

Ideally, I want to do the following;
1) Share a form with my students where they type in their name, email address and choose a LLM they want API access for.
2) I upload the spreadsheet to the gateway (or manually enter it, doesn't matter) and for each user I generate an API key that I can share with them.
3) They use the API key generated in 2) above and make API calls to portkey gateway which then redirects them to the appropriate LLM (LLM to provider will be mapped).

So basically I can generate "n" API keys for my students and have all these "n" keys mapped to 1 Anyscale API key.

Is this the way virtual-keys work? Because the virtual key I created is a modified version of the virtual key name. And it doesn't seem like this is the key my student can use to make API calls. Do correct me if I've misunderstood.

**Comment by vrushankportkey** [comment]
Hi @harshsinghal yes, this is how virtual keys work on the hosted platfrom! 

<img width="536" alt="image" src="https://github.com/Portkey-AI/gateway/assets/134934501/0572b32a-97ce-45e5-85d6-3d275282f87a">

You can put the same Anyscale API key above and put different virtual key names. This way, you'd have created multiple virtual keys that all map to the same, single Anyscale API key.

Currently, we auto generate the virtual key and its name for you:

<img width="1000" alt="image" src="https://github.com/Portkey-AI/gateway/assets/134934501/7772a4d5-5629-4aae-afa7-6b48bfc586d4">

You can just copy the "key" from above and share it with your students → They can then use it in their apps directly. [This doc](https://portkey.ai/docs/product/ai-gateway-streamline-llm-integrations/virtual-keys) explains how. 

If you are making the REST API requests on `https://api.portkey.ai/v1/chat/completions` route for example, you can pass the virtual key with the `'x-portkey-virtual-key: student-name-1-4969ed'` header. With this then, you'll NOT NEED to pass the `'Authorization: Bearer xxx'` header for Anyscale.

Hope this helps!

**Comment by ayush-portkey** [comment]
@harshsinghal Does the solution shared by @vrushankportkey solve your problem?

**Comment by harshsinghal** [comment]
I'm yet to test this out and will update once I have. Thanks a ton for the detailed and prompt response folks.

**Comment by vrushankportkey** [comment]
@harshsinghal checking in! Please let us know if we can help anywhere!

**Comment by vrushankportkey** [comment]
Closing this for now. @harshsinghal if you have any questions, please feel free to directly email me at vrushank.v@portkey.ai

**Comment by aqeeb-warraich** [comment]
Per-key quota for a cohort (students, in your case) is a narrower version of a problem we went deep on. Enforcement before the call, not reporting after. Here's how it works: https://unitsense.ai/governance. Happy to share how we handle per-key budget enforcement if it's still a gap for you
